/*
 * Copyright contributors to Hyperledger Besu.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
package org.hyperledger.besu.sila.sil.messages;

import static org.assertj.core.api.Assertions.assertThat;

import org.hyperledger.besu.datatypes.Hash;
import org.hyperledger.besu.sila.core.BlockDataGenerator;
import org.hyperledger.besu.sila.p2p.rlpx.wire.MessageData;
import org.hyperledger.besu.sila.p2p.rlpx.wire.RawMessage;
import org.hyperledger.besu.sila.rlp.BytesValueRLPOutput;

import java.math.BigInteger;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import org.junit.jupiter.api.Test;

public class GetBlockAccessListsMessageTest {

  @Test
  public void roundTripTest() {
    final BlockDataGenerator generator = new BlockDataGenerator(1);
    final List<Hash> blockHashes = new ArrayList<>();
    final int hashCount = 20;
    for (int i = 0; i < hashCount; i++) {
      blockHashes.add(generator.hash());
    }

    final MessageData initialMessage = GetBlockAccessListsMessage.create(blockHashes);
    final MessageData raw =
        new RawMessage(SilProtocolMessages.GET_BLOCK_ACCESS_LISTS, initialMessage.getData());
    final GetBlockAccessListsMessage message = GetBlockAccessListsMessage.readFrom(raw);

    final Iterator<Hash> readBlockHashes = message.blockHashes().iterator();
    for (int i = 0; i < hashCount; i++) {
      assertThat(readBlockHashes.next()).isEqualTo(blockHashes.get(i));
    }
    assertThat(readBlockHashes.hasNext()).isFalse();
  }

  @Test
  public void wrapsWithEth71WireShape() {
    // [request-id, [hashes]]
    final BlockDataGenerator generator = new BlockDataGenerator(1);
    final Hash blockHash = generator.hash();

    final GetBlockAccessListsMessage message =
        GetBlockAccessListsMessage.create(List.of(blockHash));
    final MessageData wrapped = message.wrapMessageData(BigInteger.valueOf(7));

    final BytesValueRLPOutput expected = new BytesValueRLPOutput();
    expected.startList();
    expected.writeBigIntegerScalar(BigInteger.valueOf(7));
    expected.startList();
    expected.writeBytes(blockHash.getBytes());
    expected.endList();
    expected.endList();

    assertThat(wrapped.getData()).isEqualTo(expected.encoded());
  }

  @Test
  public void createWithEmptyHashes() {
    final MessageData initialMessage = GetBlockAccessListsMessage.create(List.of());
    final MessageData raw =
        new RawMessage(SilProtocolMessages.GET_BLOCK_ACCESS_LISTS, initialMessage.getData());
    final GetBlockAccessListsMessage message = GetBlockAccessListsMessage.readFrom(raw);

    assertThat(message.blockHashes()).isEmpty();
  }
}
